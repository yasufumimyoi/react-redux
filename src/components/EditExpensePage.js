import React from "react";

const EditExpensePage = () => {
  return (
    <div>
      <p>Editing</p>
    </div>
  );
};

export default EditExpensePage;
