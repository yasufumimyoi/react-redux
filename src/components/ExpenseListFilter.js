import React from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
} from "../action/filters";
import { DateRangePicker } from "react-dates";
import { v4 as uuidv4 } from "uuid";

//storeから値を貰っているのと同時にdispatch関数も受け取っている
//Textの値はstoreからpropsとして持って来ている為、dispatchしないと変更する事ができないので、onChangeを使用する
//onChangeする度にdispatchして、storeに新しい値を書き込んでいる
//classコンポーネントで、他のコンポーネントからpropsを受け取った場合は、this.propsで使う

class ExpenseListFilter extends React.Component {
  state = {
    calenderFocused: null,
    startDateId: uuidv4(),
    endDateId: uuidv4(),
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };
  onFocusChange = (calenderFocused) => {
    this.setState(() => ({ calenderFocused }));
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filter.text}
          onChange={(e) => {
            this.props.dispatch(setTextFilter(e.target.value));
          }}
        />
        <select
          //最初に表示されるoptionの値をstoreから取って来ている
          //dateが初期値でoptionもdateが先に来ているので、値と表示が一致している
          value={this.props.filter.value}
          onChange={(e) => {
            if (e.target.value === "date") {
              this.props.dispatch(sortByDate());
            } else if (e.target.value === "amount") {
              this.props.dispatch(sortByAmount());
            }
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filter.startDate}
          startDateId={this.state.startDateId}
          endDate={this.props.filter.endDate}
          endDateId={this.state.endDateId}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

//storeからどの値を取ってくるか決める
const mapStateToProps = (state) => {
  return {
    filter: state.filter,
  };
};

//storeからどの値を取ってくるか決める
export default connect(mapStateToProps)(ExpenseListFilter);
