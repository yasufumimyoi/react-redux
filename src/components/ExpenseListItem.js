import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";

//ExpenseListとstoreの両方からpropsを受け取っている
const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>
      {amount} - {createdAt}
    </p>
  </div>
);

//connectに引数を与えなくてもdispatchをpropsとして渡す事が出来る
export default connect()(ExpenseListItem);
