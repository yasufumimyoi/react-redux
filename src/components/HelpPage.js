import React from "react";

const HelpPage = () => (
  <div>
    <p>This is from my help component</p>
  </div>
);

export default HelpPage;
