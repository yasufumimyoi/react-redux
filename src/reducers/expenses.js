const expensesReducerDefaultState = [];

//初期値がなければ空の配列が使われる
//初期値がStoreに既にあれば、その値が初期値が配列の中にオブジェクトで配置される
//Actionの値(オブジェクト)を引数名actionとして受け取っている
//ReducerはPure function
//outputはinputのみに依存している事
const expensesReducers = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      //配列の中でstateオブジェクトを回している
      //その後で新しいオブジェクトを追加している
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => {
        return id !== action.id;
      });
    case "EDIT_EXPENSE":
      //編集したいオブジェクトを特定する為にIDを渡す
      //expenseオブジェクトの全てのキーと値を回す
      //最後にupdatesで渡した値で上書して編集する
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates,
          };
        } else {
          return state;
        }
      });

    default:
      return state;
  }
};

export default expensesReducers;
