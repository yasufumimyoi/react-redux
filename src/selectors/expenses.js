//stateを引数として受け取ってソートする役目
//元データの値とソートするのに必要な値を渡して、処理した値をreturnで返して来ている
//第一引数はexpense(配列の中にオブジェクト)で第二引数はfilter(オブジェクト)

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const startDateMatch =
        //createdAtはExpenseの値 startDateはFilterの値
        //Filterをかける時の値がstartDateなので、この値が基準になる
        //startDateよりも大きい値が残る
        typeof startDate !== "number" || expense.createdAt >= startDate;
      const endDateMatch =
        //createdAtはExpenseの値 startDateはFilterの値
        //endDateよりも値が小さいものが残る
        typeof endDate !== "number" || expense.createdAt <= endDate;
      const textMatch = expense.description
        //expenseのdescriptionとfilterのtextの両方を小文字にする必要がある
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};

export default getVisibleExpenses;
