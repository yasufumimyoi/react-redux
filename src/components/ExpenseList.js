import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import getVisibleExpenses from "../selectors/expenses";

//このpropsはstoreから来ている
//配列の中にオブジェクト型のexpensesのstateが渡されている
const ExpenseList = (props) => {
  return (
    <div>
      <h2>ExpenseList</h2>
      {props.expenses.map((expense) => {
        return (
          <ExpenseListItem
            key={expense.id}
            {...expense}
            // spreadを使えば以下のような書き方をしなくても済む
            // description={expense.description}
            // amount={expense.amount}
            // createdAt={expense.createdAt}
            // key={index.toString}
          />
        );
      })}
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filter),
  };
};

//mapStateToPropsはstateを戻す
//dispatchされて、storeのstateがupdateされたらrenderされる値も変更する
//よってライフサイクルを使う必要がなくなる
export default connect(mapStateToProps)(ExpenseList);

//stateをstoreから持って来ている
//その値をリターンしてExpenseListに渡して上げている
//それをpropsとして受け取って使用

// const ConnectedExpenseList = connect((state) => {
//   return {
//     expenses: state.expenses,
//   };
// })(ExpenseList);

// export default ConnectedExpenseList;
