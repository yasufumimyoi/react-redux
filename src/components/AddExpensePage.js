import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startAddExpense } from "../action/expenses";

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h2>Add Expense</h2>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
