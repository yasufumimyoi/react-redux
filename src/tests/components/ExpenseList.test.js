import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpenseList";
import { ExpenseListItem } from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";
import toJSON from "enzyme-to-json";

test("should render ExpenseList with expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should render ExpenseList with no expenses props", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should render ExpenseListItem", () => {
  //expense={...expense}だとオブジェクトを渡してしまう
  //下記の通りだと最初のオブジェクトの値を渡す事ができる
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
