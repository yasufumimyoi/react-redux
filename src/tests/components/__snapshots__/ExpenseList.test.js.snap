// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <Connect(ExpenseListItem)
    amount={195}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={4500}
    createdAt={345600000}
    description="Gum"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`should render ExpenseList with no expenses props 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render ExpenseListItem 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      Gum
    </h3>
  </Link>
  <p>
    195
     - 
    0
  </p>
</div>
`;
